tests/test_batch_edit.py

```
@@ -396,3 +396,175 @@ def test_batch_operation_error_is_docx_edit_error(self, multi_para_doc):
+        assert paragraphs_after
+        assert all(hasattr(p, "parse_position") for p in paragraphs_after)
+
+    def test_rollback_failure_surfaces_original_error(self, multi_para_doc, monkeypatch):
+        """If the rollback re-parse itself fails, the original edit error must
+        still propagate — not the rollback failure."""
+        doc, _ = multi_para_doc
+        refs = doc.list_paragraphs()
+        editor = doc._document_editor
+        before_text = doc.get_visible_text()
+
+        def boom(xml_bytes):
```

Ruff (0.15.15)

[warning] Missing return type annotation for private function `boom`

Add return type annotation: `NoReturn`

(ANN202)